src/datetime.rs

```diff
@@ -339,50 +339,33 @@ impl DateTime {
-                        // fraction is either:
-                        // - up to 3 digits of millisecond fractions, i.e. microseconds
-                        // - or up to 6 digits of second fractions, i.e. milliseconds
-                        let max_digits = if timestamp > MS_WATERSHED { 3 } else { 6 };
-                        let Some(fract_integers) = int_parse_bytes(fract) else {
-                            return Err(e);
-                        };
-                        if config.microseconds_precision_overflow_behavior
-                            == MicrosecondsPrecisionOverflowBehavior::Error
-                            && fract.len() > max_digits
-                        {
-                            return Err(if timestamp > MS_WATERSHED {
-                                ParseError::MillisecondFractionTooLong
-                            } else {
-                                ParseError::SecondFractionTooLong
-                            });
+            Err(e) => match float_parse_bytes(bytes) {
+                IntFloat::Int(int) => Self::from_timestamp_with_config(int, 0, config),
+                IntFloat::Float(float) => {
+                    let timestamp_in_milliseconds = float.abs() > MS_WATERSHED as f64;
+
+                    if config.microseconds_precision_overflow_behavior == MicrosecondsPrecisionOverflowBehavior::Error {
+                        let fractional_digits = float.to_string().split('.').nth(1).unwrap_or("").len();
+
+                        if timestamp_in_milliseconds && fractional_digits > 3 {
+                            return Err(ParseError::MillisecondFractionTooLong);
+                        } else if !timestamp_in_milliseconds && fractional_digits > 6 {
+                            return Err(ParseError::SecondFractionTooLong);
                         }
-                        // TODO: Technically this is rounding, but this is what the existing
-                        // behaviour already did. Probably this is always better than "truncating"
-                        // so we might want to change MicrosecondsPrecisionOverflowBehavior and
-                        // make other uses also round / deprecate truncating.
-                        let multiple = 10f64.powf(max_digits as f64 - fract.len() as f64);
-                        Self::from_timestamp_with_config(
-                            timestamp,
-                            (fract_integers as f64 * multiple).round() as u32,
-                            config,
-                        )
                     }
+
+                    let timestamp_normalized: f64 = if timestamp_in_milliseconds {
+                        float / 1_000f64
+                    } else {
+                        float
+                    };
+                    let seconds = timestamp_normalized.floor() as i64;
+                    let microseconds = ((timestamp_normalized - seconds as f64) * 1_000_000f64).round() as u32;
```

Comment:
@latk,

Right, the issue is that we need `floor` to handle negative numbers correctly, because we only support positive microseconds (which is a whole other can of worms, imo).

Reply:
I think this merits some extra comments in the code, because it's a pretty confusing pattern.

Reply:
We could do the rough thing again where we do `.trunc()` then `-1` for `seconds` if the value is negative, and `1_000_000 - microseconds`, but I find this approach a bit more clean.

Reply:
@davidhewitt,

Any idea for a fix here re the rounding issue?

Reply:
Maybe fixed by https://github.com/pydantic/speedate/pull/71/commits/252c619087449155a76835443320f3b63db6eca6